README.md
![YMF825Arduino](resource/header.png)

## Features

- Supports two [YMF825Board](https://yamaha-webmusic.github.io/ymf825board/intro/) for stereo
- Supports manipulation of any operator
- Supports manipulation of any 3-band equalizer

## How to use

After installing the library, include `ymf825.h`.

`ymf825.h` supports low-level functions and allows direct manipulation of YMF825 registers.
`ymf825/driver.h` supports high-level functions and wraps register operations. In addition, `ymf825/tonePresets.h` includes presets for tone parameters.

```cpp
#include "ymf825.h"
#include "ymf825/driver.h"
#include "ymf825/tonePresets.h"
```

Specify reset (/RESET) and slave select (/SS) pins of the YMF825 and declare the device and driver variables.

```cpp
#define PIN_RESET PIN_A6
#define PIN_SS PIN_A7

YMF825 ymf825(PIN_RESET, PIN_SS);
YMF825Driver driver(&ymf825);
```

Initialize the YMF825 before you begin operating the device. `initialize(voltage)` function resets the device with hardware reset. The parameter must be set to correct voltage as a constant.
The YMF825 does not set any tone parameters to sound immediately after initialization. You can set any tone parameters with `setTone(tone)` function.

```cpp
void setup() {
  driver.initialize(YMF825_VOLTAGE_3V3);
  driver.setTone(YMF825TonePresets::Default);
}
```

To start the sound, call `noteOn(voice, octave, tone)` function. To stop the sound, call `noteOff(voice) ` function. As with MIDI, nothing happens when you call `noteOff` when there is no sound.

```cpp
void loop() {
  const uint8_t NOTES[] = {
    YMF825_NOTE_C, YMF825_NOTE_D, YMF825_NOTE_E, YMF825_NOTE_F,
    YMF825_NOTE_G, YMF825_NOTE_F, YMF825_NOTE_E, YMF825_NOTE_D,
  };

  for (uint8_t i = 0; i < sizeof(NOTES); i++) {
    driver.noteOff(0);  // for voice #0
    delay(25);
    driver.noteOn(0, 4, NOTES[i]);  // for voice #0, octave 4
    delay(475);
  }
}
```

See [here](./examples/simple_tone/simple_tone.ino) for a compilable sketch.

## License

[__MIT License__](./LICENSE)


ChangeLog.md
## 1.0.1

- Added a README.
- Added a change log.
- Added defines and reduced magic numbers.
  - `YMF825_VOICE_COUNT`, `YMF825_TONE_SIZE`, `YMF825_TONE_COUNT`, `YMF825_OPERATOR_SIZE`, `YMF825_OPERATOR_COUNT`, `YMF825_EQUALIZER_BAND_COUNT`, `YMF825_EQUALIZER_BAND_SIZE`
- Added examples.
  - stereo_tone.ino
- Added a macro `opAssignment` in toneParameter.cpp. This macro is used to simplify writing assignment to specified operator.
- Replaced for-loop to `memcpy`.

## 1.0 (2020/06/11)

- Published first version.


src/ymf825.h
#pragma once

#include "Arduino.h"

#define YMF825_SPI_MAX_SPEED 10000000

enum YMF825TargetChip {
  YMF825_TARGET_NONE = 0x00,
  YMF825_TARGET_LCH  = 0x01,
  YMF825_TARGET_RCH  = 0x02,

  YMF825_TARGET_ALL = YMF825_TARGET_LCH | YMF825_TARGET_RCH,

  YMF825_TARGET_MONO   = YMF825_TARGET_LCH,
  YMF825_TARGET_STEREO = YMF825_TARGET_LCH | YMF825_TARGET_RCH,
};

class YMF825Device {
 protected:
 public:
  virtual void initialize() = 0;

  virtual void resetHardware() = 0;

  virtual bool isBusy() = 0;

  virtual void setTargetChip(uint8_t targetChip) = 0;
  virtual uint8_t getTargetChip()                = 0;

  virtual void write(uint8_t address, uint8_t data)                      = 0;
  virtual void burstwrite(uint8_t address, uint8_t *data, uint16_t size) = 0;

  virtual uint8_t read(uint8_t targetChip, uint8_t address) = 0;
};

class YMF825 : public YMF825Device {
 private:
  uint8_t resetPin;
  uint8_t ssLchPin;
  uint8_t ssRchPin;
  uint8_t targetChip;

 public:
  YMF825(uint8_t resetPin, uint8_t ssPin);
  YMF825(uint8_t resetPin, uint8_t ssLchPin, uint8_t ssRchPin);

  void initialize();

  void resetHardware();

  bool isBusy();

  void setTargetChip(uint8_t targetChip);
  uint8_t getTargetChip();

  void write(uint8_t address, uint8_t data);
  void burstwrite(uint8_t address, uint8_t *data, uint16_t size);

  uint8_t read(uint8_t targetChip, uint8_t address);

 private:
  void enableSlaveSelect();
  void enableSlaveSelectExclusive(uint8_t targetChip);
  void disableSlaveSelect();
};


src/ymf825.cpp
#include "ymf825.h"

#include "SPI.h"

const SPISettings spiSettings = SPISettings(YMF825_SPI_MAX_SPEED, MSBFIRST, SPI_MODE0);

YMF825::YMF825(uint8_t resetPin, uint8_t ssPin) {
  this->resetPin = resetPin;
  this->ssLchPin = ssPin;
  this->ssRchPin = ssPin;
  this->setTargetChip(YMF825_TARGET_MONO);
}

YMF825::YMF825(uint8_t resetPin, uint8_t ssLchPin, uint8_t ssRchPin) {
  this->resetPin = resetPin;
  this->ssLchPin = ssLchPin;
  this->ssRchPin = ssRchPin;
  this->setTargetChip(YMF825_TARGET_STEREO);
}

void YMF825::initialize() {
  pinMode(resetPin, OUTPUT);
  pinMode(ssLchPin, OUTPUT);
  pinMode(ssRchPin, OUTPUT);

  digitalWrite(resetPin, HIGH);
  this->disableSlaveSelect();

  SPI.begin();
}

void YMF825::resetHardware() {
  digitalWrite(resetPin, LOW);
  delay(1);
  digitalWrite(resetPin, HIGH);
}

bool YMF825::isBusy() {
  return false;
}

void YMF825::setTargetChip(uint8_t targetChip) {
  this->targetChip = targetChip;
}

uint8_t YMF825::getTargetChip() {
  return this->targetChip;
}

void YMF825::write(uint8_t address, uint8_t data) {
  SPI.beginTransaction(spiSettings);
  {
    this->enableSlaveSelect();
    SPI.transfer(address);
    SPI.transfer(data);
    this->disableSlaveSelect();
  }
  SPI.endTransaction();
}

void YMF825::burstwrite(uint8_t address, uint8_t *data, uint16_t size) {
  SPI.beginTransaction(spiSettings);
  {
    this->enableSlaveSelect();
    SPI.transfer(address);
    SPI.transfer(data, size);
    this->disableSlaveSelect();
  }
  SPI.endTransaction();
}

uint8_t YMF825::read(uint8_t targetChip, uint8_t address) {
  uint8_t data;

  SPI.beginTransaction(spiSettings);
  {
    this->enableSlaveSelectExclusive(targetChip);
    SPI.transfer(0x80 | address);
    data = SPI.transfer(data);
    this->disableSlaveSelect();
  }
  SPI.endTransaction();

  return data;
}

void YMF825::enableSlaveSelect() {
  // LOW is active, HIGH is inactive
  digitalWrite(ssLchPin, (targetChip & YMF825_TARGET_LCH) ? LOW : HIGH);
  digitalWrite(ssRchPin, (targetChip & YMF825_TARGET_RCH) ? LOW : HIGH);
}

void YMF825::enableSlaveSelectExclusive(uint8_t targetChip) {
  digitalWrite((targetChip & YMF825_TARGET_LCH) ? ssLchPin : ssRchPin, LOW);
  digitalWrite((targetChip & YMF825_TARGET_LCH) ? ssRchPin : ssLchPin, HIGH);
}

void YMF825::disableSlaveSelect() {
  digitalWrite(ssLchPin, HIGH);
  digitalWrite(ssRchPin, HIGH);
}


src/ymf825/driver.cpp
#include "ymf825/driver.h"

const double fnumTable[12] = {
  // fnumTable[i] = getFnum(440.0 * pow(2.0, (48.0 + i - 69.0) / 12.0), 3);
  357.206105157084,    // A
  378.44668549375604,  // A#
  400.9502964632338,   // B
  424.7920417752144,   // C
  450.05149103837175,  // C#
  476.812945316528,    // D
  505.16571847561823,  // D#
  535.204435260415,    // E
  567.0293470958184,   // F
  600.7466666666667,   // F#
  636.4689223926988,   // G
  674.3153339816948,   // G#
};

YMF825Driver::YMF825Driver(YMF825Device *device) {
  this->device = device;
}

void YMF825Driver::setClockEnable(bool enable) {
  device->write(0x00, enable ? 0x01 : 0x00);
}

bool YMF825Driver::getClockEnable(uint8_t targetChip) {
  return device->read(targetChip, 0x00) == 0x01;
}

void YMF825Driver::setAllRegisterReset(bool state) {
  device->write(0x01, state ? 0x80 : 0x00);
}

bool YMF825Driver::getAllRegisterReset(uint8_t targetChip) {
  return device->read(targetChip, 0x01) == 0x80;
}

void YMF825Driver::setAnalogBlockPowerDown(uint8_t block) {
  device->write(0x02, block & 0x0f);
}

uint8_t YMF825Driver::getAnalogBlockPowerDown(uint8_t targetChip) {
  return device->read(targetChip, 0x02) & 0x0f;
}

void YMF825Driver::setGain(uint8_t gainLevel) {
  device->write(0x03, gainLevel & 0x03);
}

uint8_t YMF825Driver::getGain(uint8_t targetChip) {
  return device->read(targetChip, 0x03) & 0x03;
}

uint8_t YMF825Driver::getHardwareId(uint8_t targetChip) {
  return device->read(targetChip, 0x04) & 0x03;
}

void YMF825Driver::setToneParameters(ToneParameters *params, uint8_t *buffer) {
  buffer[0] = 0x80 + YMF825_TONE_COUNT;

  for (uint8_t i = 0; i < YMF825_TONE_COUNT; i++) {
    ToneParameter *toneParameter = params->getParameter(i);
    memcpy(buffer + YMF825_TONE_SIZE * i + 1, toneParameter->memory, YMF825_TONE_SIZE);
  }

  buffer[YMF825_TONE_SIZE * YMF825_TONE_COUNT + 1] = 0x80;
  buffer[YMF825_TONE_SIZE * YMF825_TONE_COUNT + 2] = 0x03;
  buffer[YMF825_TONE_SIZE * YMF825_TONE_COUNT + 3] = 0x81;
  buffer[YMF825_TONE_SIZE * YMF825_TONE_COUNT + 4] = 0x80;

  setToneParameter(buffer, YMF825_TONE_SIZE * YMF825_TONE_COUNT + 5);
}

void YMF825Driver::setToneParameter(ToneParameter *param) {
  uint8_t buffer[YMF825_TONE_SIZE + 5];

  buffer[0] = 0x81;
  memcpy(buffer + 1, param->memory, YMF825_TONE_SIZE);
  buffer[YMF825_TONE_SIZE + 1] = 0x80;
  buffer[YMF825_TONE_SIZE + 2] = 0x03;
  buffer[YMF825_TONE_SIZE + 3] = 0x81;
  buffer[YMF825_TONE_SIZE + 4] = 0x80;

  setToneParameter(buffer, sizeof(buffer));
}

void YMF825Driver::setToneParameter(const uint8_t *preset) {
  ToneParameter toneParameter(preset);
  setToneParameter(&toneParameter);
}

void YMF825Driver::setToneParameter(uint8_t *data, uint16_t size) {
  device->burstwrite(0x07, data, size);
}

void YMF825Driver::setSequencerSetting(uint8_t sequencerSetting) {
  device->write(0x08, sequencerSetting);
}

uint8_t YMF825Driver::getSequencerSetting(uint8_t targetChip) {
  return device->read(targetChip, 0x08);
}

void YMF825Driver::setSequencerVolume(uint8_t volume, bool applyInterpolation, uint16_t size) {
  volume &= 0x1f;
  size &= 0x01ff;
  device->write(0x09, (uint8_t)(volume << 3 | (applyInterpolation ? 0x00 : 0x04) | size >> 8));
  device->write(0x0a, (uint8_t)(size & 0xff));
}

void YMF825Driver::getSequencerVolume(uint8_t targetChip, uint8_t *volume, bool *applyInterpolation, uint16_t *size) {
  uint8_t msb = device->read(targetChip, 0x09);
  uint8_t lsb = device->read(targetChip, 0x0a);

  *volume             = msb >> 3;
  *applyInterpolation = (msb & 0x04) == 0;
  *size               = ((uint16_t)msb & 0x01) << 8 | lsb;
}

void YMF825Driver::setVoiceNumber(uint8_t number) {
  device->write(0x0b, number & 0x0f);
}

uint8_t YMF825Driver::getVoiceNumber(uint8_t targetChip) {
  return device->read(targetChip, 0x0b) & 0x0f;
}

void YMF825Driver::setVoiceVolume(uint8_t volume) {
  device->write(0x0c, (volume & 0x1f) << 2);
}

void YMF825Driver::setFnumAndBlock(uint16_t fnum, uint8_t block) {
  fnum &= 0x03ff;
  block &= 0x07;

  device->write(0x0d, ((fnum & 0x0380) >> 4) | (block & 0x07));
  device->write(0x0e, fnum & 0x7f);
}

void YMF825Driver::setToneFlag(uint8_t toneNumber, bool keyOn, bool mute, bool resetEnvelopeGenerator) {
  device->write(0x0f, ((keyOn ? 0x40 : 0x00) | (mute ? 0x20 : 0x00) | (resetEnvelopeGenerator ? 0x10 : 0x00) |
                       toneNumber & 0x0f));
}

void YMF825Driver::setChannelVolume(uint8_t volume, bool applyInterpolation) {
  device->write(0x10, (volume & 0x1f) << 2 | (applyInterpolation ? 0x00 : 0x01));
}

void YMF825Driver::setVibrationModuration(uint8_t depth) {
  device->write(0x11, depth & 0x07);
}

void YMF825Driver::setFrequencyMultiplier(uint8_t integer, uint16_t fraction) {
  fraction &= 0x01ff;
  device->write(0x12, (uint8_t)((integer & 0x03) << 3 | fraction >> 6));
  device->write(0x13, (uint8_t)((fraction & 0x3f) << 1));
}

void YMF825Driver::setInterporationInMuteState(bool enable) {
  device->write(0x14, enable ? 0x00 : 0x01);
}

void YMF825Driver::setSequencerTimeUnitSetting(uint16_t timeUnit) {
  timeUnit &= 0x3fff;
  device->write(0x17, (uint8_t)(timeUnit >> 7));
  device->write(0x18, (uint8_t)(timeUnit & 0x7f));
}

void YMF825Driver::setMasterVolume(uint8_t volume) {
  device->write(0x19, (uint8_t)((volume & 0x3f) << 2));
}

uint8_t YMF825Driver::getMasterVolume(uint8_t targetChip) {
  return device->read(targetChip, 0x19) >> 2;
}

void YMF825Driver::setSoftResetFlag(uint8_t value) {
  device->write(0x1a, value);
}

uint8_t YMF825Driver::getSoftResetFlag(uint8_t targetChip) {
  return device->read(targetChip, 0x1a);
}

void YMF825Driver::setVolumeInterpolationSetting(bool dadjt, uint8_t muteItime, uint8_t chvolItime, uint8_t mvolItime) {
  device->write(0x1b,
                ((dadjt ? 0x40 : 0x00) | ((muteItime & 0x03) << 4) | ((chvolItime & 0x03) << 2) | (mvolItime & 0x03)));
}

void YMF825Driver::getVolumeInterpolationSetting(uint8_t targetChip, bool *dadjt, uint8_t *muteItime,
                                                 uint8_t *chvolItime, uint8_t *mvolItime) {
  uint8_t readByte = device->read(targetChip, 0x1b);
  *dadjt           = (readByte & 0x40) != 0;
  *muteItime       = (readByte >> 4) & 0x03;
  *chvolItime      = (readByte >> 2) & 0x03;
  *mvolItime       = readByte & 0x03;
}

void YMF825Driver::setLfoResetFlag(bool reset) {
  device->write(0x1c, reset ? 0x01 : 0x00);
}

bool YMF825Driver::getLfoResetFlag(uint8_t targetChip) {
  return device->read(targetChip, 0x1c) == 0x01;
}

void YMF825Driver::setPowerRailSelection(uint8_t voltage) {
  device->write(0x1d, voltage & 0x01);
}

uint8_t YMF825Driver::getPowerRailSelection(uint8_t targetChip) {
  return device->read(targetChip, 0x1d) & 0x01;
}

void YMF825Driver::setEqualizer(uint8_t band, const double *coefficients) {
  uint8_t data[YMF825_EQUALIZER_BAND_SIZE * YMF825_EQUALIZER_BAND_COUNT];

  for (uint8_t i = 0; i < YMF825_EQUALIZER_BAND_SIZE; i++) {
    uint32_t registerFormat = YMF825Driver::serializeCoefficient(coefficients[i]);

    data[i * YMF825_EQUALIZER_BAND_COUNT + 0] = (uint8_t)(registerFormat >> 16);
    data[i * YMF825_EQUALIZER_BAND_COUNT + 1] = (uint8_t)((registerFormat >> 8) & 0xff);
    data[i * YMF825_EQUALIZER_BAND_COUNT + 2] = (uint8_t)(registerFormat & 0xff);
  }

  device->burstwrite(0x20 + band, data, sizeof(data));
}

uint32_t YMF825Driver::getEqualizerCoefficient(uint8_t targetChip, uint8_t band, uint8_t coefficientNumber) {
  uint8_t baseAddress = 0x23 + band * YMF825_EQUALIZER_BAND_SIZE * YMF825_EQUALIZER_BAND_COUNT +
                        coefficientNumber * YMF825_EQUALIZER_BAND_COUNT;
  uint8_t byte0 = device->read(targetChip, baseAddress + 0);
  uint8_t byte1 = device->read(targetChip, baseAddress + 1);
  uint8_t byte2 = device->read(targetChip, baseAddress + 2);
  return (byte0 << 16) + (byte1 << 8) + byte2;
}

void YMF825Driver::getEqualizer(uint8_t targetChip, uint8_t band, double *coefficients) {
  for (uint8_t i = 0; i < YMF825_EQUALIZER_BAND_SIZE; i++)
    coefficients[i] = YMF825Driver::deserializeCoefficient(getEqualizerCoefficient(targetChip, band, i));
}

void YMF825Driver::setSoftwareTestCommunication(uint8_t value) {
  device->write(0x50, value);
}

uint8_t YMF825Driver::getSoftwareTestCommunication(uint8_t targetChip) {
  return device->read(targetChip, 0x50);
}

// Static Functions

double YMF825Driver::getFnum(double frequency, uint8_t block) {
  // Original Formula
  // const fnum = (freq * Math.pow(2, 19)) / (Math.pow(2, block - 1) * 48000);
  return pow(2.0, 13 - block) * frequency / 375.0;
}

double YMF825Driver::getFrequency(double fnum, uint8_t block) {
  // Original Formula
  // const freq = (48000 * Math.pow(2, block - 1) * fnum) / Math.pow(2, 19);
  return 375.0 * pow(2.0, block - 13) * fnum;
}

void YMF825Driver::getFnumAndBlock(uint8_t key, double *fnum, uint8_t *block, double *correction) {
  if (key < 0 || key > 127)
    key = 0;  // error

  double blockMod = 1.0;
  *block          = key / 12 - 2;

  if (*block < 0) {
    blockMod = pow(2.0, *block);
    *block   = 0;
  } else if (*block > 6) {
    blockMod = pow(2.0, *block - 6);
    *block   = 6;
  }

  *fnum            = fnumTable[key % 12] * blockMod;
  double idealFreq = getFrequency(*fnum, *block);

  if (*fnum > 1023.0)
    *fnum = 1023.0;

  *correction = idealFreq / getFrequency(round(*fnum), *block);
}

void YMF825Driver::getFraction(double multiplier, uint8_t *integer, uint16_t *fraction) {
  if (multiplier >= 4.0 || multiplier < 0.0)
    multiplier = 0.0;  // error

  multiplier = round(multiplier * 512.0);

  if (multiplier >= 4 * 512) {
    *integer  = 3;
    *fraction = 511;
  } else {
    *integer  = (uint8_t)(multiplier / 512.0);
    *fraction = (uint16_t)(multiplier - *integer * 512);
  }
}

uint32_t YMF825Driver::serializeCoefficient(double coefficient) {
  if (isnan(coefficient) || isinf(coefficient) || coefficient <= -8.0 || coefficient >= 8.0)
    coefficient = 0.0;  // error

  uint8_t integer   = (uint8_t)abs(coefficient);
  uint32_t fraction = (uint32_t)round(abs(coefficient - (uint8_t)coefficient) * 1048575.0);

  // 2'complement
  if (coefficient < 0.0) {
    integer  = 0x08 | (~integer & 0x07);
    fraction = ~fraction & 0x0fffff;
  }

  // Sign bit:       1 bit  (CEQ##[23])
  // Integer part:   3 bits (CEQ##[22:20]) -> 2'complement
  // Fraction part: 20 bits (CEQ##[19:0])  -> 2'complement
  return (integer << 20) | fraction;
}

double YMF825Driver::deserializeCoefficient(uint32_t registerFormat) {
  if (registerFormat >> 24 > 0)
    registerFormat = 0.0;  // error

  // value is negative when sign is true
  bool sign         = (registerFormat & 0x00800000) != 0;
  uint8_t integer   = (registerFormat & 0x00700000) >> 20;
  uint32_t fraction = registerFormat & 0x000fffff;

  // 2'complement
  if (sign) {
    integer  = ~integer & 0x07;
    fraction = ~fraction & 0x0fffff;
  }

  return (integer + fraction / 1048575.0) * (sign ? -1.0 : 1.0);
}

// High-level Functions

void YMF825Driver::initialize(uint8_t voltage) {
  device->initialize();
  device->setTargetChip(YMF825_TARGET_ALL);
  resetHardware();
  resetSoftware(voltage);

  for (uint8_t i = 0; i < YMF825_VOICE_COUNT; i++) {
    setVoiceNumber(i);
    setVoiceVolume(0x10);
  }
}

void YMF825Driver::resetHardware() {
  device->resetHardware();
}

void YMF825Driver::resetSoftware(uint8_t voltage) {
  setPowerRailSelection(voltage);
  setAnalogBlockPowerDown(0x0e);
  delay(1);
  setClockEnable(true);
  setAllRegisterReset(false);
  setSoftResetFlag(0xa3);
  delay(1);
  setSoftResetFlag(0x00);
  delay(30);
  setAnalogBlockPowerDown(0x04);
  delay(1);
  setAnalogBlockPowerDown(0x00);

  setMasterVolume(0x3c);
  setVolumeInterpolationSetting(false, 0x03, 0x03, 0x03);
  setInterporationInMuteState(true);
  setGain(0x01);

  setSequencerSetting(0xf6);
  delay(21);
  setSequencerSetting(0x00);
  setSequencerVolume(0x1f, true, 0x0000);

  setSequencerTimeUnitSetting(0x2000);
}

void YMF825Driver::setToneAssignment(uint8_t voice, uint8_t tone) {
  setVoiceNumber(voice);
  device->write(0x0f, toneFlag[voice] = (toneFlag[voice] & 0xf0) | (tone & 0x0f));
}

void YMF825Driver::noteOn(uint8_t voice, uint8_t key) {
  double fnum, correction;
  uint8_t block, integer;
  uint16_t fraction;

  getFnumAndBlock(key, &fnum, &block, &correction);
  getFraction(correction, &integer, &fraction);

  setVoiceNumber(voice);
  setFrequencyMultiplier(integer, fraction);
  setFnumAndBlock((uint16_t)fnum, block);
  device->write(0x0f, toneFlag[voice] |= 0b01000000);
}

void YMF825Driver::noteOn(uint8_t voice, int8_t octave, uint8_t key) {
  noteOn(voice, (octave + 2) * 12 + key - 3);
}

void YMF825Driver::noteOff(uint8_t voice) {
  setVoiceNumber(voice);
  device->write(0x0f, toneFlag[voice] &= 0b10111111);
}

void YMF825Driver::setTone(ToneParameters *params, uint8_t *buffer) {
  setSequencerSetting(0x16);
  setSequencerSetting(0x00);
  setToneParameters(params, buffer);
}

void YMF825Driver::setTone(ToneParameter *param) {
  setSequencerSetting(0x16);
  setSequencerSetting(0x00);
  setToneParameter(param);
}

void YMF825Driver::setTone(const uint8_t *preset) {
  setSequencerSetting(0x16);
  setSequencerSetting(0x00);
  setToneParameter(preset);
}

void YMF825Driver::setTone(uint8_t *data, uint16_t size) {
  setSequencerSetting(0x16);
  setSequencerSetting(0x00);
  setToneParameter(data, size);
}


src/ymf825/driver.h
#pragma once

#include "Arduino.h"
#include "ymf825.h"

#define YMF825_VOICE_COUNT 16
#define YMF825_TONE_SIZE 30
#define YMF825_TONE_COUNT 16
#define YMF825_OPERATOR_SIZE 7
#define YMF825_OPERATOR_COUNT 4

#define YMF825_EQUALIZER_BAND_COUNT 3
#define YMF825_EQUALIZER_BAND_SIZE 5

enum YMF825_VOLTAGE {
  YMF825_VOLTAGE_5V  = 0,
  YMF825_VOLTAGE_3V3 = 1,
};

enum YMF825_NOTE {
  YMF825_NOTE_A = 0,
  YMF825_NOTE_B = 2,
  YMF825_NOTE_C = 3,
  YMF825_NOTE_D = 5,
  YMF825_NOTE_E = 7,
  YMF825_NOTE_F = 8,
  YMF825_NOTE_G = 10,
};

class ToneParameter {
 public:
  uint8_t memory[YMF825_TONE_SIZE] = { 0 };

  ToneParameter();
  ToneParameter(const uint8_t *source);

  void fillMemory(const uint8_t *source);

  uint8_t getBaseOctave();
  void setBaseOctave(uint8_t value);
  uint8_t getLfoFrequency();
  void setLfoFrequency(uint8_t value);
  uint8_t getAlgorithm();
  void setAlgorithm(uint8_t value);

  uint8_t getSustainRate(uint8_t operatorNumber);
  void setSustainRate(uint8_t operatorNumber, uint8_t value);
  bool getIgnoreKeyOff(uint8_t operatorNumber);
  void setIgnoreKeyOff(uint8_t operatorNumber, bool value);
  bool getEnableKeyScaleSensitivity(uint8_t operatorNumber);
  void setEnableKeyScaleSensitivity(uint8_t operatorNumber, bool value);
  uint8_t getReleaseRate(uint8_t operatorNumber);
  void setReleaseRate(uint8_t operatorNumber, uint8_t value);
  uint8_t getDecayRate(uint8_t operatorNumber);
  void setDecayRate(uint8_t operatorNumber, uint8_t value);
  uint8_t getAttackRate(uint8_t operatorNumber);
  void setAttackRate(uint8_t operatorNumber, uint8_t value);
  uint8_t getSustainLevel(uint8_t operatorNumber);
  void setSustainLevel(uint8_t operatorNumber, uint8_t value);
  uint8_t getTotalLevel(uint8_t operatorNumber);
  void setTotalLevel(uint8_t operatorNumber, uint8_t value);
  uint8_t getKeyScalingLevel(uint8_t operatorNumber);
  void setKeyScalingLevel(uint8_t operatorNumber, uint8_t value);

  uint8_t getAmplitudeModurationDepth(uint8_t operatorNumber);
  void setAmplitudeModurationDepth(uint8_t operatorNumber, uint8_t value);
  bool getEnableAmplitudeModuration(uint8_t operatorNumber);
  void setEnableAmplitudeModuration(uint8_t operatorNumber, bool value);
  uint8_t getVibrationDepth(uint8_t operatorNumber);
  void setVibrationDepth(uint8_t operatorNumber, uint8_t value);
  bool getEnableVibration(uint8_t operatorNumber);
  void setEnableVibration(uint8_t operatorNumber, bool value);

  uint8_t getMagnificationOfFrequency(uint8_t operatorNumber);
  void setMagnificationOfFrequency(uint8_t operatorNumber, uint8_t value);
  uint8_t getDetune(uint8_t operatorNumber);
  void setDetune(uint8_t operatorNumber, uint8_t value);

  uint8_t getWaveShape(uint8_t operatorNumber);
  void setWaveShape(uint8_t operatorNumber, uint8_t value);
  uint8_t getFeedbackLevel(uint8_t operatorNumber);
  void setFeedbackLevel(uint8_t operatorNumber, uint8_t value);
};

class ToneParameters {
 public:
  ToneParameter parameters[YMF825_TONE_COUNT] = { ToneParameter() };

  void setParameter(uint8_t toneNumber, ToneParameter *parameter);
  void setParameter(uint8_t toneNumber, const uint8_t *buffer);
  ToneParameter *getParameter(uint8_t toneNumber);
};

class YMF825Driver {
 private:
  YMF825Device *device;
  uint8_t toneFlag[YMF825_VOICE_COUNT] = { 0 };

 public:
  YMF825Driver(YMF825Device *device);

  // Low-level functions

  void setClockEnable(bool enable);
  bool getClockEnable(uint8_t targetChip);

  void setAllRegisterReset(bool state);
  bool getAllRegisterReset(uint8_t targetChip);

  void setAnalogBlockPowerDown(uint8_t block);
  uint8_t getAnalogBlockPowerDown(uint8_t targetChip);

  void setGain(uint8_t gainLevel);
  uint8_t getGain(uint8_t targetChip);

  uint8_t getHardwareId(uint8_t targetChip);

  void setToneParameters(ToneParameters *params, uint8_t *buffer);

  void setToneParameter(ToneParameter *param);
  void setToneParameter(const uint8_t *preset);
  void setToneParameter(uint8_t *data, uint16_t size);

  void setSequencerSetting(uint8_t sequencerSetting);
  uint8_t getSequencerSetting(uint8_t targetChip);

  void setSequencerVolume(uint8_t volume, bool applyInterpolation, uint16_t size);
  void getSequencerVolume(uint8_t targetChip, uint8_t *volume, bool *applyInterpolation, uint16_t *size);

  void setVoiceNumber(uint8_t number);
  uint8_t getVoiceNumber(uint8_t targetChip);

  void setVoiceVolume(uint8_t volume);

  void setFnumAndBlock(uint16_t fnum, uint8_t block);

  void setToneFlag(uint8_t toneNumber, bool keyOn, bool mute, bool resetEnvelopeGenerator);

  void setChannelVolume(uint8_t volume, bool applyInterpolation);

  void setVibrationModuration(uint8_t depth);

  void setFrequencyMultiplier(uint8_t integer, uint16_t fraction);

  void setInterporationInMuteState(bool enable);

  void setSequencerTimeUnitSetting(uint16_t timeUnit);

  void setMasterVolume(uint8_t volume);
  uint8_t getMasterVolume(uint8_t targetChip);

  void setSoftResetFlag(uint8_t value);
  uint8_t getSoftResetFlag(uint8_t targetChip);

  void setVolumeInterpolationSetting(bool dadjt, uint8_t muteItime, uint8_t chvolItime, uint8_t mvolItime);
  void getVolumeInterpolationSetting(uint8_t targetChip, bool *dadjt, uint8_t *muteItime, uint8_t *chvolItime,
                                     uint8_t *mvolItime);

  void setLfoResetFlag(bool reset);
  bool getLfoResetFlag(uint8_t targetChip);

  void setPowerRailSelection(uint8_t select);
  uint8_t getPowerRailSelection(uint8_t targetChip);

  void setEqualizer(uint8_t band, const double *coefficients);
  uint32_t getEqualizerCoefficient(uint8_t targetChip, uint8_t band, uint8_t coefficientNumber);
  void getEqualizer(uint8_t targetChip, uint8_t band, double *coefficients);

  void setSoftwareTestCommunication(uint8_t value);
  uint8_t getSoftwareTestCommunication(uint8_t targetChip);

  // Static Functions

  static double getFnum(double frequency, uint8_t block);
  static double getFrequency(double fnum, uint8_t block);
  static void getFnumAndBlock(uint8_t key, double *fnum, uint8_t *block, double *correction);
  static void getFraction(double multiplier, uint8_t *integer, uint16_t *fraction);

  static uint32_t serializeCoefficient(double coefficient);
  static double deserializeCoefficient(uint32_t registerFormat);

  // High-level Functions

  void resetHardware();
  void resetSoftware(uint8_t voltage);

  void initialize(uint8_t voltage);
  void setToneAssignment(uint8_t voice, uint8_t tone);

  void noteOn(uint8_t voice, uint8_t key);
  void noteOn(uint8_t voice, int8_t octave, uint8_t key);

  void noteOff(uint8_t voice);

  void setTone(ToneParameters *params, uint8_t *buffer);
  void setTone(ToneParameter *param);
  void setTone(const uint8_t *preset);
  void setTone(uint8_t *data, uint16_t size);
};
